# Analysing the sc2ts ARG using tskit, numpy and pandas

This notebooks shows examples of analysing the 3.13 million sample sc2ts ARG using the [tskit Python API](https://tskit.dev/tskit/docs/), in conjuction with Pandas and with an emphasis on using vectorised approaches with Numpy. This is to showcase the ease and efficiency with which pandemic scale analyses can be done directly within a notebook environment using only Python, and to provide a starting point for further analyses. We focus on the core computational tasks one might like to perform on a pandemic scale ARG, with the minimal set of external dependencies required to do this.

Please see the tskit [documentation](https://tskit.dev/tskit/docs/stable/) for detailed information on the Python APIs, and the  [tskit tutorials](https://tskit.dev/tutorials/intro.html) site for an introduction to tskit and help with various tasks such as visualisation. Please see the [preprint](https://www.biorxiv.org/content/10.1101/2023.06.08.544212) for background and further details on all aspects of sc2ts and the ARG.

In [4]:
import tszip
import sc2ts
import numpy as np

First, we load the ARG into memory from the compressed representation using [tszip](https://tskit.dev/software/tszip.html), which typically takes less than a second. The object we get back is a [tskit TreeSequence](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence), which we call ``ts`` by convention.

In [5]:
ts = tszip.load("../data/sc2ts_viridian_v2.0.trees.tsz")

When viewed in a notebook environment, the TreeSequence gives some summary information about its contents:

In [6]:
ts

## Information on nodes

The [sc2ts Python API](https://tskit.dev/sc2ts/docs/stable/api.html) provides a few utility functions that take advantage of specific structure of the sc2ts ARG and its metadata. The ``node_data`` function returns a Pandas Dataframe summarising the nodes in the ARG, and in particular, makes the key ``sample_id``, ``pango`` and ``scorpio`` metadata fields efficiently accessible.

In [7]:
dfn = sc2ts.node_data(ts)
dfn

,pango,sample_id,scorpio,node_id,is_sample,is_recombinant,num_mutations,date
0,B,Vestigial_ignore,.,0,False,False,0,2019-11-19
1,B,Wuhan/Hu-1/2019,.,1,False,False,0,2019-12-26
2,A,SRR11772659,.,2,True,False,1,2020-01-19
3,B,SRR11397727,.,3,True,False,0,2020-01-24
4,B,SRR11397730,.,4,True,False,0,2020-01-24
...,...,...,...,...,...,...,...,...
3424141,HV.1.1,,Omicron (XBB.1-like),3424141,False,False,3,2023-09-09
3424142,BQ.1.1.29,,Omicron (BA.5-like),3424142,False,False,9,2022-10-02
3424143,BA.1.1,,Omicron (BA.1-like),3424143,False,False,1,2021-11-02
3424144,AY.75,,Delta (B.1.617.2-like),3424144,False,False,9,2021-05-17


Pandas provides many powerful functions. For example, here is a breakdown of the numbers of nodes by their Scorpio assignment:

In [8]:
dfn["scorpio"].value_counts()

scorpio
Delta (B.1.617.2-like)           710451
Delta (AY.4-like)                610495
Omicron (BA.2-like)              510003
Omicron (BA.1-like)              463152
Alpha (B.1.1.7-like)             334938
Omicron (BA.5-like)              276878
.                                203951
Delta (AY.4.2-like)              100680
Omicron (XBB.1.5-like)            64955
Omicron (XBB.1-like)              44559
Omicron (BA.4-like)               33788
Omicron (XBB.1.16-like)           15110
Iota (B.1.526-like)               11802
Epsilon (B.1.429-like)             9168
Omicron (XBB-like)                 8864
Gamma (P.1-like)                   6966
Epsilon (B.1.427-like)             4842
Beta (B.1.351-like)                4127
Mu (B.1.621-like)                  1592
Omicron (XE-like)                  1311
Delta (B.1.617.2-like) +K417N      1298
B.1.1.318-like                     1211
Eta (B.1.525-like)                 1010
B.1.617.1-like                      572
Probable Omicron (Unassigned)   

We can then use this to identify specific nodes. Suppose we wanted to find the first Alpha node. To do this, we simply need to find oldest node with the corresponding Pango label.

**NOTE** The dataframe returned by ``node_data`` has a ``date`` field, which gives the estimated date for each node, which has day-precision. In order to find the oldest node reliably we need finer precision, which is provided by the ``time`` column in the tskit 
[node table](https://tskit.dev/tskit/docs/stable/data-model.html#node-table). To access this information in Pandas, we add this field to the dataframe. Then, we sort the rows by time and then extract the first record for each Pango designation:

In [9]:
dfn["time"] = ts.nodes_time
dfn_pango = dfn.sort_values("time", ascending=False).groupby(["pango"]).first()
dfn_pango

,sample_id,scorpio,node_id,is_sample,is_recombinant,num_mutations,date,time
pango,,,,,,,,
A,,.,9,False,False,1,2019-12-26,1624.000000
A.1,,.,279,False,False,2,2020-01-19,1600.000000
A.11,SRR23586010,.,4099,True,False,3,2020-03-28,1531.000000
A.2,,.,663,False,False,4,2020-02-01,1587.000000
A.2.2,ERR12537611,.,1163,True,False,1,2020-03-19,1540.000000
...,...,...,...,...,...,...,...,...
XW,,Omicron (BA.2-like),1378600,False,True,1,2022-03-03,826.540427
XY,,Omicron (Unassigned),1417358,False,True,2,2022-03-15,814.000000
XZ,,Omicron (BA.2-like),1383518,False,False,1,2022-03-09,820.000000


We can then extract the "alpha origin" node easily:

In [10]:
alpha_origin = dfn_pango.loc["B.1.1.7"]
alpha_origin

sample_id                             
scorpio           Alpha (B.1.1.7-like)
node_id                          94157
is_sample                        False
is_recombinant                   False
num_mutations                       36
date               2020-09-16 00:00:00
time                       1359.164372
Name: B.1.1.7, dtype: object

## Information on mutations

The sc2ts Python API also provides a way to extract information about mutations. 

In [11]:
dfm = sc2ts.mutation_data(ts).set_index("node")
dfm

,mutation_id,site_id,position,parent,inherited_state,derived_state,date
node,,,,,,,
9044,0,0,56,-1,G,T,2020-03-22
664838,1,0,56,-1,G,T,2021-07-19
479708,2,0,56,-1,G,T,2021-07-31
549470,3,0,56,-1,G,T,2021-09-01
802956,4,0,56,-1,G,T,2021-09-07
...,...,...,...,...,...,...,...
1850758,2761029,29639,29803,-1,C,T,2023-09-13
1868897,2761030,29639,29803,-1,C,T,2023-10-30
1875756,2761031,29639,29803,-1,C,T,2023-11-17


This dataframe has information about each mutation, including the genome position (``position``) and node over which the mutation occured. Here, we're interested in finding all the mutations immediately ancestral to the Alpha origin node found above, and so we add an index on the ``node`` field. Then, getting the mutations is quick and easy:

In [12]:
alpha_origin_mutations = dfm.loc[alpha_origin["node_id"]]
alpha_origin_mutations

,mutation_id,site_id,position,parent,inherited_state,derived_state,date
node,,,,,,,
94157,203213,854,913,-1,C,T,2020-05-23
94157,601221,5315,5388,-1,C,A,2020-05-23
94157,645961,5910,5986,-1,C,T,2020-05-23
94157,733254,6874,6954,-1,T,C,2020-05-23
94157,1035770,11192,11288,-1,T,-,2020-05-23
94157,1035822,11193,11289,-1,C,-,2020-05-23
94157,1035962,11194,11290,-1,T,-,2020-05-23
94157,1036056,11195,11291,-1,G,-,2020-05-23
94157,1036150,11196,11292,-1,G,-,2020-05-23


Here we can see there's quite a lot of mutations (36), but many of them have the derived state of the gap character ("-"). These represent the approximation of deletions in sc2ts, in which deletions at each site are modelled indepenently. 

We can extract the actual deletion events from this representation by looking at runs of mutations to the gap character and extracting their start and length. We do this using some pandas group-by logic.

In [13]:
dels = alpha_origin_mutations[alpha_origin_mutations["derived_state"] == "-"].copy()
dels["run"] = (dels["position"].diff(1) != 1).cumsum()
dels

,mutation_id,site_id,position,parent,inherited_state,derived_state,date,run
node,,,,,,,,
94157,1035770,11192,11288,-1,T,-,2020-05-23,1
94157,1035822,11193,11289,-1,C,-,2020-05-23,1
94157,1035962,11194,11290,-1,T,-,2020-05-23,1
94157,1036056,11195,11291,-1,G,-,2020-05-23,1
94157,1036150,11196,11292,-1,G,-,2020-05-23,1
94157,1036309,11197,11293,-1,T,-,2020-05-23,1
94157,1036444,11198,11294,-1,T,-,2020-05-23,1
94157,1036577,11199,11295,-1,T,-,2020-05-23,1
94157,1036686,11200,11296,-1,T,-,2020-05-23,1


In [14]:
dels_grp = dels.groupby("run")
summary = dels_grp.first() 
summary["length"] = dels_grp.last()["position"] - summary["position"] + 1
summary[["position", "length"]]

,position,length
run,,
1,11288,9
2,21765,6
3,21991,3
4,28271,1


So, we have 4 deletions: a 9 bp deletion starting at position 11288, 6bp at 21765, etc.

Non deletions are more straightforward to analyse, but it's important to factor out the deletions first because of their multibase nature. This is easily done:

In [15]:
non_dels = alpha_origin_mutations[alpha_origin_mutations["derived_state"] != "-"]
non_dels

,mutation_id,site_id,position,parent,inherited_state,derived_state,date
node,,,,,,,
94157,203213,854,913,-1,C,T,2020-05-23
94157,601221,5315,5388,-1,C,A,2020-05-23
94157,645961,5910,5986,-1,C,T,2020-05-23
94157,733254,6874,6954,-1,T,C,2020-05-23
94157,1262113,14568,14676,-1,C,T,2020-05-23
94157,1297273,15169,15279,-1,C,T,2020-05-23
94157,1349148,16063,16176,-1,T,C,2020-05-23
94157,1931984,22922,23063,-1,A,T,2020-05-23
94157,1946320,23130,23271,-1,C,A,2020-05-23


## Recombination nodes and local trees

The ``node_data`` dataframe has a field ``is_recombinant`` which makes finding the recombination nodes in the sc2ts ARG easy

In [16]:
df_recomb = dfn[dfn.is_recombinant]
df_recomb

,pango,sample_id,scorpio,node_id,is_sample,is_recombinant,num_mutations,date,time
1844,B.1.157,,.,1844,False,True,2,2020-03-18,1541.000000
25908,B.1.1,,.,25908,False,True,2,2020-07-05,1432.152172
30404,B.1.160,,.,30404,False,True,1,2020-07-03,1434.066261
31804,B.1.426,,.,31804,False,True,0,2020-08-28,1378.000000
36136,B.1.221,,.,36136,False,True,0,2020-06-27,1440.911496
...,...,...,...,...,...,...,...,...,...
1893637,JN.1,,Omicron (BA.2-like),1893637,False,True,3,2024-03-31,67.580387
1893977,JN.1.4.3,,Omicron (BA.2-like),1893977,False,True,2,2024-04-06,61.256651
1894585,XDV.1,,Omicron (BA.2-like),1894585,False,True,8,2023-11-29,190.395686
1894803,JN.1.7.5,,Omicron (BA.2-like),1894803,False,True,2,2024-05-06,31.263776


XBB is an important recombinant and is well captured by the sc2ts ARG. Its "origin node" is a recombination node, meaning that it has more than one parent in the ARG. We can see that the ``is_recombinant`` flag is True:

In [17]:
xbb_origin = dfn_pango.loc["XBB"]
xbb_origin

sample_id                            
scorpio            Omicron (XBB-like)
node_id                       1676682
is_sample                       False
is_recombinant                   True
num_mutations                      15
date              2022-05-25 00:00:00
time                       743.161421
Name: XBB, dtype: object

To find where the breakpoint is (almost all recombinants in the current sc2ts ARG have only 2 parents and consequently one breakpoint), we must inspect the tskit [edge table](https://tskit.dev/tskit/docs/stable/data-model.html#edge-table). This simply records the IDs of the parent and child nodes and the interval over which this relationship applies. We then find the edges which record the parents of the XBB origin by looking for those edges in which it is a child:

In [18]:
xbb_edges = np.where(ts.edges_child == xbb_origin["node_id"])[0]
xbb_edges

array([513777, 692291])

We get two edge IDs, which we can then use to get the properties of the edges themselves. One quick way to visualise the information is to access the edge table directly:

In [19]:
ts.tables.edges[xbb_edges]

id,left,right,parent,child,metadata
0,"22,577","29,904","1,631,432","1,676,682",
1,0,"22,577","1,306,761","1,676,682",


We can also look at the parents of the event directly:

In [20]:
dfn.loc[ts.edges_parent[xbb_edges]]

,pango,sample_id,scorpio,node_id,is_sample,is_recombinant,num_mutations,date,time
1631432,BM.1.1.1,,Omicron (BA.2-like),1631432,False,False,1,2022-04-18,780.837695
1306761,BA.2.10,ERR9201084,Omicron (BA.2-like),1306761,True,False,1,2022-02-26,831.000000


To get the breakpoint, we can then take the maximum of the left coordinate:

(Note, we're using two different conventions for accessing the underlying table data here --- they are essentially equivalent and can be mixed and matched as is most convenient.)

In [21]:
bp = np.max(ts.edges_left[xbb_edges])
bp

np.float64(22577.0)

The *effect* of this recombination event is to change the tree to the left and right of the breakpoint. Tskit provides an efficient view over the "local trees" along the genome that result from recombination events. In this case, we would like to examine the trees to the right and left of the breakpoint for XBB. The simplest (and most efficient) way to do this is to first look at the right-hand tree using the ``ts.at`` method. 

In [22]:
tree = ts.at(bp)
tree

The tree returned here is an instance of the tskit [Tree](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.Tree) class which provides many operations. This is built around the efficient numerical representation of trees, and specifically, the 
[quintuply linked tree](https://tskit.dev/tskit/docs/stable/data-model.html#quintuply-linked-trees) data structure, in which the tree topology is encoded using a set of numpy arrays. For example, the parent of each node in the tree is represented using the ``parent_array``:

In [23]:
tree.parent_array

array([     -1,      -1,       9, ..., 3421907, 1797765,      -1],
      shape=(3424147,), dtype=int32)

So, ``tree.parent_array[u]`` means "what is the node ID of the parent of node ``u`` in this tree". The ``-1`` values here indicate that a particular node has no parent, and so is either a root or is not "present" in the current tree. We can then verify that the local tree has the correct parent for the right-hand side of the XBB recombination event:

In [24]:
tree.parent_array[xbb_origin["node_id"]]

np.int32(1631432)

We can use the same approach to access the tree on the left hand side of the breakpoint, but it is more efficient to "mutate" the current tree instance using the ``prev()`` method. This is because ``ts.at()`` requires us to create a new tree instance and to apply all the edges that intersect with the given point. Using the ``prev`` method, on the other hand, takes advantage of the fact that there's very little difference between the trees at either side of the breakpoint and makes those minimal changes.

In [25]:
tree.prev()
tree

Now we can also verify that the left-hand parent of the XBB recombination node is what we extracted from the node table above.

In [26]:
tree.parent_array[xbb_origin["node_id"]]

np.int32(1306761)

## Getting the ancestors of a node in a tree

An operation that we're often interested in is to get the ancestors of a node in a particular tree. Using the ``ancestors`` methods along with Pandas selection provides an easy way to access data on the ancestors of the Alpha origin node.

In [27]:
tree = ts.first()   
u = alpha_origin["node_id"]
dfn.iloc[[u] + list(tree.ancestors(u))]    

,pango,sample_id,scorpio,node_id,is_sample,is_recombinant,num_mutations,date,time
94157,B.1.1.7,,Alpha (B.1.1.7-like),94157,False,False,36,2020-09-16,1359.164372
5396,B.1.1,,.,5396,False,False,3,2020-01-28,1591.003346
118,B.1.1,,.,118,False,False,3,2020-01-28,1591.004716
67,B.1,,.,67,False,False,1,2020-01-28,1591.515332
12,B.1,SRR11597205,.,12,True,False,2,2020-01-28,1591.515332
33,B,,.,33,False,False,1,2019-12-26,1624.000000
1,B,Wuhan/Hu-1/2019,.,1,False,False,0,2019-12-26,1624.000000


**Note:** In this example we use the first tree in the sequence to access the ancestors of the Alpha origin node. This is valid here because there is no recombinant among those ancestors. When recombination is present among the ancestors, more care is required.

## Iterating over all trees

A common pattern in analysing ARGs with tskit is to iterate sequentially over all local trees along the genome. This is done using the ``.trees()`` iterator, which again takes advantage of the fact that adjacent trees are very similar for efficiency. Here we use this to count the number of polytomies (nodes with > 2 children) in each local tree. This is done using Numpy functions on the ``num_children_array``, which is a zero-copy view on a per-node count maintained by the underlying C library.

In [28]:
%%time
num_polytomies = np.zeros(ts.num_trees, dtype=int)
for tree in ts.trees():
    num_polytomies[tree.index] = np.sum(tree.num_children_array > 2)
np.min(num_polytomies), np.max(num_polytomies)

CPU times: user 1.59 s, sys: 60 ms, total: 1.65 s
Wall time: 1.64 s


(np.int64(272637), np.int64(272675))

## Examining children

The numpy ``argmax`` function gives us a conveient way to find the ID of the node which has the largest number of children:

In [29]:
tree = ts.first()
u = np.argmax(tree.num_children_array)
tree.num_children_array[u]

np.int32(13876)

This has an impressive 13,876 children! We can first look at its path back to root, and verify that it is valid across all trees by checking that there are no recombinants in its ancestry:

In [30]:
dfn.loc[[u] + list(tree.ancestors(u))]

,pango,sample_id,scorpio,node_id,is_sample,is_recombinant,num_mutations,date,time
960619,BA.2.5,,Omicron (BA.2-like),960619,False,False,1,2021-11-27,922.010023
990529,BA.2,,Omicron (BA.2-like),990529,False,False,1,2021-11-27,922.010023
1037770,BA.2,,Omicron (BA.2-like),1037770,False,False,2,2021-11-27,922.010023
940471,BA.2,SRR17461792,Omicron (BA.2-like),940471,True,False,37,2021-11-27,922.010023
959864,BA.2,,Omicron (Unassigned),959864,False,False,0,2020-12-12,1272.053355
940472,BA.2,,Omicron (Unassigned),940472,False,False,24,2020-12-04,1280.531711
1896655,B.1.1.529,,Probable Omicron (Unassigned),1896655,False,False,56,2020-05-06,1492.502163
866911,B.1.1,,.,866911,False,False,1,2020-01-28,1591.003671
5904,B.1.1,,.,5904,False,False,1,2020-01-28,1591.003671
118,B.1.1,,.,118,False,False,3,2020-01-28,1591.004716


We can then get the list of children for that node using the ``tree.children()`` method, and extract the relevant rows from the node data table:

In [32]:
df_children = dfn.loc[list(tree.children(u))]
df_children

,pango,sample_id,scorpio,node_id,is_sample,is_recombinant,num_mutations,date,time
960617,BA.2,ERR12895802,Omicron (BA.2-like),960617,True,False,1,2021-12-01,918.000000
960618,BA.2,SRR18518948,Omicron (BA.2-like),960618,True,False,1,2021-12-01,918.000000
1040187,BA.2.10,SRR18518942,Omicron (BA.2-like),1040187,True,False,1,2021-12-18,901.000000
1051564,BA.2,SRR17888957,Omicron (BA.2-like),1051564,True,False,1,2021-12-22,897.000000
1051565,BA.2,SRR18518984,Omicron (BA.2-like),1051565,True,False,1,2021-12-22,897.000000
...,...,...,...,...,...,...,...,...,...
3414491,BA.2,,Omicron (BA.2-like),3414491,False,False,1,2022-01-04,884.847330
3418604,BA.2,,Omicron (BA.2-like),3418604,False,False,2,2021-12-22,897.485794
3419540,BA.2,,Omicron (BA.2-like),3419540,False,False,6,2022-02-09,848.525598
3421631,BA.2,,Omicron (BA.2-like),3421631,False,False,4,2021-12-06,913.718034


When we break down the counts of those nodes by their sample status and number of mutations, we see that we have over 10,000 samples which are identical to the focal node. Thus, this seems like a well-inferred node capturing a very deeply sampled phase of the BA.2 outbreak.

In [33]:
df_children[["num_mutations", "is_sample"]].value_counts()

num_mutations  is_sample
0              True         10241
1              True          2459
               False          729
2              True           305
3              True            64
2              False           53
4              True            14
3              False            4
5              True             3
4              False            2
9              True             1
6              False            1
Name: count, dtype: int64

## All descendants

Getting the descendants of a node in a particular tree is straightforward, and can be done efficiently in the standard traversal orders. For example, here we get all the nodes descending from the Alpha origin in the first tree in preorder, and extract the corresponding dataframe:


In [34]:
tree = ts.first()
descendants = tree.preorder(alpha_origin["node_id"])
descendants

array([ 94157,  50504,  50473, ..., 299925, 301786,  93676],
      shape=(335498,), dtype=int32)

In [35]:
alpha_descendants = dfn.loc[descendants]
alpha_descendants

,pango,sample_id,scorpio,node_id,is_sample,is_recombinant,num_mutations,date,time
94157,B.1.1.7,,Alpha (B.1.1.7-like),94157,False,False,36,2020-09-16,1359.164372
50504,B.1.1.7,,Alpha (B.1.1.7-like),50504,False,False,1,2020-09-16,1359.164372
50473,B.1.1.7,ERR4827009,Alpha (B.1.1.7-like),50473,True,False,1,2020-10-21,1324.000000
56801,B.1.1.7,ERR4862251,Alpha (B.1.1.7-like),56801,True,False,1,2020-11-05,1309.000000
61508,B.1.1.7,ERR4905821,Alpha (B.1.1.7-like),61508,True,False,1,2020-11-11,1303.000000
...,...,...,...,...,...,...,...,...,...
3421867,B.1.1.7,,Alpha (B.1.1.7-like),3421867,False,False,1,2021-01-03,1250.712755
218583,B.1.1.7,SRR16617144,Alpha (B.1.1.7-like),218583,True,False,2,2021-03-24,1170.000000
299925,B.1.1.7,ERR5999240,Alpha (B.1.1.7-like),299925,True,False,6,2021-05-24,1109.000000
301786,B.1.1.7,ERR6010769,Alpha (B.1.1.7-like),301786,True,False,1,2021-05-26,1107.000000


However, this is an approximation to the true descendants of the node, because there are some recombinants descending from it. 

In [36]:
alpha_descendants[alpha_descendants["is_recombinant"]]

,pango,sample_id,scorpio,node_id,is_sample,is_recombinant,num_mutations,date,time
303512,B.1.1.7,,Alpha (B.1.1.7-like),303512,False,True,0,2021-05-12,1121.000000
287930,B.1.526,,Iota (B.1.526-like),287930,False,True,2,2021-04-21,1142.934617
273929,B.1.1.7,,Alpha (B.1.1.7-like),273929,False,True,3,2021-04-11,1152.585251
366791,Q.3,,Alpha (B.1.1.7-like),366791,False,True,1,2021-06-26,1076.395052
185098,B.1.1.7,,Alpha (B.1.1.7-like),185098,False,True,3,2021-03-03,1191.000000
216781,B.1.1.7,,Alpha (B.1.1.7-like),216781,False,True,1,2021-03-15,1179.000000
258158,B.1.1.7,,Alpha (B.1.1.7-like),258158,False,True,1,2021-04-01,1162.525500
316027,B.1.1.7,,Alpha (B.1.1.7-like),316027,False,True,1,2021-05-13,1120.073620
277112,B.1.526,,Iota (B.1.526-like),277112,False,True,1,2021-04-13,1150.216038
259283,B.1.1.7,,Alpha (B.1.1.7-like),259283,False,True,0,2021-03-01,1193.000000


To get the true set of descendants of this node we must currently take the union of the descendants across all trees as follows.

**NOTE:** efficient support for this operation is planned in tskit

In [37]:
all_descendants = []
for tree in ts.trees():
    all_descendants = np.union1d(
        all_descendants, np.sort(tree.preorder(alpha_origin["node_id"])))

In [38]:
alpha_descendants = dfn.loc[all_descendants]
alpha_descendants

,pango,sample_id,scorpio,node_id,is_sample,is_recombinant,num_mutations,date,time
50473,B.1.1.7,ERR4827009,Alpha (B.1.1.7-like),50473,True,False,1,2020-10-21,1324.000000
50474,B.1.1.7,ERR4827181,Alpha (B.1.1.7-like),50474,True,False,0,2020-10-21,1324.000000
50475,B.1.1.7,ERR4833995,Alpha (B.1.1.7-like),50475,True,False,1,2020-10-21,1324.000000
50476,B.1.1.7,ERR4827287,Alpha (B.1.1.7-like),50476,True,False,1,2020-10-22,1323.000000
50477,B.1.1.7,ERR4848983,Alpha (B.1.1.7-like),50477,True,False,0,2020-10-23,1322.000000
...,...,...,...,...,...,...,...,...,...
3424043,B.1.1.7,,Alpha (B.1.1.7-like),3424043,False,False,1,2021-03-14,1180.000000
3424110,B.1.1.7,,Alpha (B.1.1.7-like),3424110,False,False,1,2021-01-14,1239.366824
3424128,B.1.1.7,,Alpha (B.1.1.7-like),3424128,False,False,2,2021-02-24,1198.000000
3424136,B.1.1.7,,Alpha (B.1.1.7-like),3424136,False,False,2,2020-10-17,1328.713799
